server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -289,26 +293,79 @@ static class Compatible extends FiltersAggregator {
+
             return new LeafBucketCollectorBase(sub, null) {
                 @Override
                 public void collect(int doc, long bucket) throws IOException {
                     boolean matched = false;
-                    for (int i = 0; i < docFilters.length; i++) {
-                        if (docFilters[i].test(doc)) {
-                            collectBucket(sub, doc, bucketOrd(bucket, i));
-                            matched = true;
+                    if (filterIterators.size() > 0) {
+                        // Advance filters if necessary. Filters will already be advanced if used as a competitive iterator.
+                        DisiWrapper top = filterIterators.top();
+                        while (top.doc < doc) {
+                            top.doc = top.approximation.advance(doc);
+                            top = filterIterators.updateTop();
+                        }
+
+                        if (top.doc == doc) {
+                            for (DisiWrapper w = filterIterators.topList(); w != null; w = w.next) {
+                                // It would be nice if DisiPriorityQueue supported generics to avoid unchecked casts.
+                                FilterMatchingDisiWrapper topMatch = (FilterMatchingDisiWrapper) w;
+
+                                // We need to cache the result of twoPhaseView.matches() since it's illegal to call it multiple times on the
+                                // same doc, yet LeafBucketCollector#collect may be called multiple times with the same doc and multiple
+                                // buckets.
+                                if (topMatch.lastCheckedDoc < doc) {
+                                    topMatch.lastCheckedDoc = doc;
+                                    if (topMatch.twoPhaseView == null || topMatch.twoPhaseView.matches()) {
+                                        topMatch.lastMatchingDoc = doc;
+                                    }
+                                }
+                                if (topMatch.lastMatchingDoc == doc) {
+                                    collectBucket(sub, doc, bucketOrd(bucket, topMatch.filterOrd));
+                                    matched = true;
+                                }
+                            }
                         }
                     }
+
                     if (otherBucketKey != null && false == matched) {
-                        collectBucket(sub, doc, bucketOrd(bucket, docFilters.length));
+                        collectBucket(sub, doc, bucketOrd(bucket, numFilters));
+                    }
+                }
+
+                @Override
+                public DocIdSetIterator competitiveIterator() throws IOException {
+                    if (useCompetitiveIterator) {
+                        // A DocIdSetIterator view of the filterIterators heap
+                        return new DisjunctionDISIApproximation(filterIterators);
```

Comment:
Looking at the code of `DisjunctionDISIApproximation`, it will not like if `filterIterators` is empty. Maybe check for emptiness and return a `DocIdSetIterator.empty()` in that case? It should be possible to write a test case by using filters on terms that do not exist.

Reply:
Right, yeah that's why I was returning a scorer with an empty DISI. I updated the code to capture this.

Note that, if filterIterators is empty, we won't be using the competitive iterator in the latest version.